napari/_qt/qt_main_window.py

```diff
@@ -84,6 +87,7 @@ class _QtMainWindow(QMainWindow):
     def __init__(self, viewer: 'Viewer', parent=None) -> None:
         super().__init__(parent)
         self._ev = None
+        self._window = viewer.window
```

Comment:
turns out this is the reason your injection is failing.  at this point, `viewer.window` is still None.  This is why I did that ugly workaround in #4826 to add a new item to the init.  this whole window init is a big ugly mess, so let me come up with a slightly more attractive pattern in #4862 that we can use here.  (you can copy it from there for now if you want, but it will change)

Reply:
ok thanks! I think I will park here for now and will add your changes once they are ready. I can also work with others to get the other few menus migrated after we have a good pattern.

Reply:
Not sure what pattern you are planning to go with, but I would LOVE to get around some of the circular import/init with dependency injections. This may be a slightly larger refactor. But I think this can be a really important foundation for future work. 

I briefly looked into the feasibility of it, and it seems we may need a more powerful framework if we want to get rid of our super nested dependencies. I dream that eventually, no class constructs another class, and every class just declares their ctor with typed/named dependencies.

Reply:
> but I would LOVE to get around some of the circular import/init with dependency injections

yeah, you and me both :joy: 
definitely a much bigger refactor.  something I've looked into many times, using dependency graphs and stuff.  it's definitely important, but a huge task